source/particle/world.cc

```diff
@@ -0,0 +1,889 @@
+              try
+                {
+                  cell = (GridTools::find_active_cell_around_point<> (this->get_mapping(), this->get_triangulation(), lost_particle->second.get_location())).first;
+                }
+              catch (GridTools::ExcPointNotFound<dim> &)
+                {
+                  // If we can find no cell for this particle there is no hope left
+                  // to find its cell. Simply delete the particle.
+                  continue;
+                }
+
+              // Reinsert the particle into our domain if we found its cell
+              // Mark it for MPI transfer otherwise
+              if (cell->is_locally_owned())
+                {
+                  const types::LevelInd found_cell = std::make_pair(cell->level(),cell->index());
+                  moved_particles_cell.insert(std::make_pair(found_cell, lost_particle->second));
+                }
+              else
+                moved_particles_domain.insert(std::make_pair(cell->subdomain_id(),lost_particle->second));
+            }
+        }
+    }
+
+    template <int dim>
+    void
+    World<dim>::send_recv_particles(const std::multimap<types::subdomain_id,Particle <dim> > &send_particles)
+    {
+      // Determine the communication pattern
+#if DEAL_II_VERSION_GTE(8,4,0)
+      const std::vector<types::subdomain_id> neighbors (this->get_triangulation().ghost_owners().begin(),
+                                                        this->get_triangulation().ghost_owners().end());
+#else
+      const std::set<types::subdomain_id> ghost_owner = ghost_owners(this->get_triangulation());
+      const std::vector<types::subdomain_id> neighbors (ghost_owner.begin(),
+                                                        ghost_owner.end());
+#endif
+      const unsigned int num_neighbors = neighbors.size();
+      const unsigned int particle_size = property_manager->get_particle_size() + integrator->get_data_size();
+
+      // Determine the amount of data we will send to other processors
+      std::vector<int> num_send_data(num_neighbors);
+      std::vector<int> num_recv_data(num_neighbors);
+
+      std::vector<int> send_offsets(num_neighbors);
+      std::vector<int> recv_offsets(num_neighbors);
+
+      // Allocate space for sending and receiving particle data
+      std::vector<char> send_data(send_particles.size() * particle_size);
+      void *data = static_cast<void *> (&send_data.front());
+
+      int total_send_data = 0;
+      for (types::subdomain_id neighbor_id = 0; neighbor_id < num_neighbors; ++neighbor_id)
+        {
+          send_offsets[neighbor_id] = total_send_data;
+
+          std::pair< const typename std::multimap<types::subdomain_id,Particle <dim> >::const_iterator,
+              const typename std::multimap<types::subdomain_id,Particle <dim> >::const_iterator>
+              send_particle_range = send_particles.equal_range(neighbors[neighbor_id]);
```

Comment:
I guess this is the place where you're going to skip over all particles that ended up on cells for which we didn't know the subdomain_id. Can you add a comment here to this effect, in essence saying that you're abandoning these particles.